### ISSUE TITLE
[NGC Container] [JIT mode only] Nvidia nsight runtime error 

### ISSUE THREAD

**Issue by SrivastavaKshitij** [pytorch/pytorch#26438]
## 🐛 Bug
 
Hi there, I encounter runtime error when I try to profile pytorch models. Error is seen only in `jit mode` and not in `eager mode`.  I have a feeling this is somewhat related to #24891 

## Error Summary:
```
RuntimeError: 
error in dlopen or dlsym: libthnvrtc.so: cannot open shared object file: No such file or directory:
```

## Environment

`Nvidia NGC Container [19.06-py3]`

## To Reproduce

Run the script with following command line options to reproduce error:
 
```
nsys profile -w true -t 'cudnn,cuda,osrt,nvtx' -o test python nsight_jit_problem.py --jit
```

#### Script (nsight_jit_problem.py)

```
import torch
import argparse
import torchvision.models as models
import torch.backends.cudnn as cudnn
import timeit
import numpy as np
import torch.nn as nn

def printStats(graphName, timings, batch_size):
    times = np.array(timings)
    steps = len(times)
    speeds = batch_size / times
    time_mean = np.mean(times)
    time_med = np.median(times)
    time_99th = np.percentile(times, 99)
    time_std = np.std(times, ddof=0)
    speed_mean = np.mean(speeds)
    speed_med = np.median(speeds)

    msg = ("\n%s =================================\n"
            "batch size=%d, num iterations=%d\n"
            "  Median FPS: %.1f, mean: %.1f\n"
            "  Median latency: %.6f, mean: %.6f, 99th_p: %.6f, std_dev: %.6f\n"
            ) % (graphName,
                batch_size, steps,
                speed_med, speed_mean,
                time_med, time_mean, time_99th, time_std)
    print(msg)


if __name__ == '__main__':

    parser = argparse.ArgumentParser(description="Run inference on a model with random input values")
    parser.add_argument('--model', default='resnet50', type=str, help='model architecture: resnet18, resnet34, resnet50, resnet101, mobilenet, mobilnetv2')
    parser.add_argument('--jit', action='store_true', help='run jit mode')
    parser.add_argument('--gpu', default=0, type=int, help='GPU id to use.')
    parser.add_argument('--fp16', action='store_true', help='Run inference with half precision float')
    parser.add_argument("--batch_size", type=int, default=1, help="Batch size (default=1)")
    parser.add_argument("--iter", default=100, type=int, help="Number of iteration loops")
    args = parser.parse_args()

    if args.model:
        model_switch = {
            "resnet18" : models.resnet18(),
            "resnet34" : models.resnet34(),
            "resnet50" : models.resnet50(),
            "resnet101" : models.resnet101(),
            "resnet152" : models.resnet152(),
            "densenet" : models.DenseNet(),
        }
        # Creating model with random weights
        print("Creating model '{}'".format(args.model))

        model = model_switch.get(args.model, "Invalid")
        if model == "Invalid":
            raise Exception("Invalid model")
    
    modules=list(model.children())[:-1]
    model=nn.Sequential(*modules)
    model = model.eval()

    if args.jit:
        print("Generate traced model...")
        example_input = torch.rand(args.batch_size, 3, 224, 224, dtype=torch.float)
        model = torch.jit.trace(model, example_input,
            check_trace=True,
            check_tolerance=1e-05,
            optimize=True,
            )

    # Create graph on GPU if CUDA is available
    if torch.cuda.is_available():
        # Select GPU to use
        device = torch.device('cuda', args.gpu)
        # Print out some status
        print("Cuda DNN version=", torch.backends.cudnn.version())
        print("Current GPU id, GPU count=", torch.cuda.current_device(), torch.cuda.device_count())
        print("Cuda compute capability=", torch.cuda.get_device_capability(device))
        print("Cuda device name=", torch.cuda.get_device_name(device))
        # Enable CuDNN autotune for better performance (with fixed inputs)
        cudnn.benchmark = True
    else:
        # raise Exception("No cuda available.")
        device = torch.device('cpu')

    # Set input shape (NCWH)
    input_shape = (args.batch_size, 3, 224, 224)
    # Create random input tensor of certain size
    torch.manual_seed(12345)
    input_t = torch.rand(input_shape, dtype=torch.float)

    if args.fp16:
        # Cast model and data to half precision
        model = model.half()
        input_t = input_t.half()

    # Copy model and input to device
    model = model.to(device)
    input_t = input_t.to(device)
    timings=[]
    with torch.no_grad():
        for i in range(args.iter):
            start_time = timeit.default_timer()
            features = model(input_t)
            torch.cuda.synchronize()
            end_time = timeit.default_timer()
            if i ==0:
                continue
            else:
                timings.append(end_time - start_time)
            #print("Iteration {}: {:.6f} s".format(i, end_time - start_time))
    
    print("Input shape:", input_t.size())
    print("Output features size:", features.size())

    printStats(args.model, timings, args.batch_size)
```

## Full Error message

```
root@abe79ca9d148:/workspace/work/profiling_standard_benchmarks# nsys profile -w true -t 'cudnn,cuda,osrt,nvtx' -o test python nsight_jit_problem.py --jit 

**** collection configuration ****
        output_filename = test
        stop-on-exit = true
        delay = 0 seconds
        duration = 0 seconds
        inherit-environment = true
        show-output = true
        trace-fork-before-exec = false
        sample_cpu = true
        backtrace_method = LBR
        trace_cublas = false
        trace_cuda = true
        trace_cudnn = true
        trace_nvtx = true
        trace_openacc = false
        trace_vulkan = false
        trace_opengl = false
        trace_osrt = true
        osrt-threshold = 0 nanoseconds
        profile_processes = tree
        application command = python
        application arguments = nsight_jit_problem.py --jit --fp16 
        application working directory = /workspace/work/profiling_standard_benchmarks
        environment variables:
        Collecting data...
Creating model 'resnet50'
Generate traced model...
Cuda DNN version= 7600
Current GPU id, GPU count= 0 1
Cuda compute capability= (7, 5)
Cuda device name= GeForce RTX 2080
Traceback (most recent call last):
  File "nsight_jit_problem.py", line 104, in <module>
    features = model(input_t)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 494, in __call__
    result = self.forward(*input, **kwargs)
RuntimeError: 
error in dlopen or dlsym: libthnvrtc.so: cannot open shared object file: No such file or directory:
operation failed in interpreter:

        Generating the test.qdstrm file.
        Capturing raw events...
        89423 total events collected.
        Saving diagnostics...
        Saving qdstrm file to disk...
        Finished saving file.


Importing the qdstrm file using /usr/local/cuda-10.1/NsightSystems-cli-2019.3.1/Host-x86_64/QdstrmImporter.

Importing...

Importing [==================================================100%]
Saving report to file "/workspace/work/profiling_standard_benchmarks/test.qdrep"
Report file saved.
Please discard the qdstrm file and use the qdrep file instead.

Removed test.qdstrm as it was successfully imported.
Please use the qdrep file instead.

```

cc @suo

**Comment by ngimel** [comment]
This is an nsys bug, not a pytorch bug. @ptrblck or @mcarilli should be able to tell you which nsys versions have a fix.

**Comment by mcarilli** [comment]
Nsight Systems is being updated pretty frequently, so using [most recent version](https://developer.nvidia.com/nsight-systems) (2019.5.x) is the first thing I would try.

It's only a collection of libraries and binaries, so you should be able to unpack the directory, launch the container with the directory mounted in, set the PATH to point to the nsys executable, and rerun the application.  

More recent containers may not have more recent versions of nsys, because (I think) the version of nsys that's seen by default in the container is what's packaged with Cuda, and Cuda doesn't get updated that often.

**Comment by mcarilli** [comment]
I asked around and it appears that for old versions of nsys, when an application under nsys tried to dynamically load a library (eg libthnvrtc.so) on-demand (eg via dlopen) nsys intercepted the dlopen call, but nsys itself was not picking up any rpath information baked into the application itself, so nsys's intercepted dlopen call might not know where to look for libthnvrtc.so.  At least that was my rudimentary interpretation.

tl;dr Latest nsys fixes this bug, so I hear.  Exporting LD_LIBRARY_PATH to explicitly include libthnvrtc.so may also fix the issue even for older nsys, because then nsys will know where to look even if it isn't properly pulling rpath information from the application.

**Comment by SrivastavaKshitij** [comment]
thanks for the update @mcarilli : I have a question. When I untar latest Nsight system, I can only `nsys` binary under `/NVIDIA_Nsight_Systems_Linux_2019.5.1.58/opt/nvidia/nsight-systems/2019.5.1/bin` but I dont see `nv-nsight-cu-cli`. Is that packaged separately ?

**Comment by SrivastavaKshitij** [comment]
Also, inside my Pytorch `19.06-py3` container, I tried exporting path by:

`export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/conda/lib/python3.6/site-packages/torch/lib/libthnvrtc.so`

```
root@e828edc78219:/workspace/work# echo $LD_LIBRARY_PATH
/usr/local/cuda/compat/lib:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/opt/conda/lib/python3.6/site-packages/torch/lib/libthnvrtc.so
```

And I still get the same error.

**Comment by mcarilli** [comment]
`nv-nsight-cu-cli` is the cli for Nsight Compute, the new individual-kernel profiling tool.  Nsight Compute is a [completely separate product](https://developer.nvidia.com/nsight-compute-2019_4). 

The system profiler executable `nsys` looks like it's at its usual location (`target-linux-x64/nsys`) in the untarred `nsight-systems-2019.5.1` directory.  You don't need Nsight Compute to use Nsight Systems, or vice versa.

**Comment by SrivastavaKshitij** [comment]
Sorry my bad: 
I had to use `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/conda/lib/python3.6/site-packages/torch/lib/`.  Remove `libthnvrtc.so` from the path. 

Thanks for the quick response @mcarilli . Closing this issue now. :tada:
